Added maxBodySize option which required middleware factories

quantumew opened the pull request:
Hello!

We recently ran into a situation where we needed to limit the request body size. This of course required options be passed to readBody which required a factory of some kind. For consistency, I wrapped all middleware in factories. `restiq.mw` still exposes direct access to the middleware.

The changes I made enumerated:
* Wrapped all middleware in factories following a consistent naming convention. Ex. `buildReadBody`.
* restiq.mw.readBody enforces a max request size which defaults to 100kb.
* Added a dependency on bytes to parse strings given as maxBodySize. I point this out because this may be a point of contention. Perhaps we should just require a number that represents the number of bytes.
* Updated and added unit tests.
* This is somewhat unrelated but we thought it would be helpful to have a discardBody middleware. I replaced two uses of readBody with discardBody because they were only being used to consume the entire request body to force the end event to fire. Please correct me if this was improper usage.
* Finally I updated the docs.

The end goal of all of this was so we could enforce a max body size but it sort of snow balled. No rush on this pull request and feel free to disagree with any of the solutions provided.

Thanks!!

quantumew committed:
Added maxBodySize option which required me to make factories for middleware.

I updated tests, added new tests for new test cases, updated the docs,
and added bytes for parsing the maxBodySize if it is a string such as 100kb. An
addition unrelated change was I added a discardBody middleware and used
it to ensure the body gets consumed if an error occurs or it simply
wasn't read.

The end goal of all of this was so we could enforce a max body size.

quantumew commented:
Hello @andrasq, Let me know if you want me to split this pull request up or do anything different off the bat. We started using your module for a work project after some frustration with restify, please let us know if you have any intentions on reviewing this, if not we can take a different approach. Also, if you don't like the way I did something feel free to ask for changes, of course! Thank you sir!!

andrasq commented:
Hi!  I just noticed your PR by pure chance, I hadn't seen an emailed notification for it.  Give me a bit to read through the changes and comment, but in principle what you did sounds good.  As to the
specific points (thanks for the summary list, it really helps):

- I'm ok with factories, not sure if they're all needed but like you said, consistency
- I would prefer the defaut be the "same as before", ie `maxBodySize: Infinity` unlimited.  This would introduce a subtle gotcha otherwise; there is no need to make this a breakin change.  The actual limit is application dependent, eg for some of our apps tens of megabytes is not unusual
- I think I favor just plain bytes.  Clever interfaces are cute, but each such feature is a custom programming language, and the number of little syntaxes that one has to track can get unwieldy.  (Parsing is pretty simple, actually, it can be one-liner RegExp to identify a trailing `/\s*([BKMGTP]i?[B]?)/i$` and a multiplier mapping `{ B:1, b:1, K:1024, Ki:1024, k:1000, M:1024*1024, Mi:1024*1024, m:1e6, }` etc.  Ok, a 3-liner.  Still extra complexity.)
- `discardBody` is nice to have because it's descriptive, and better describes what the code is doing.  The implementation should fold the two cases, but having a distinct entry point is good.

I hope to get to this in the next day or two, and thanks!!

andrasq commented on Readme.md:
```diff
@@ -304,12 +305,61 @@ Gather up the message that was sent with the http request, and save it in
 
+### restiq.mw.discardBody( req, res, next )
+
+Reads and discards request body forcing the end event on the request.
```
"forcing" -> "to force'
maybe quote 'end' or backtick `end` to make the context more obvious

quantumew replied:
I skimmed this so I missed the back tick request. That ended up in a  separate commit 4928e3801cd304e85d86d0081f3ca610d426c738

andrasq commented on Readme.md:
```diff
@@ -280,7 +280,8 @@ params can be gotten with `app.mw.parseQueryParams()`.
 
-A library of pre-written middleware utility functions.
+A library of pre-written middleware utility functions. All middleware is
+additionally exposed as a factory.
```
maybe "... also exposed through a configurable factory function."

quantumew replied:
Much better. Fixed

andrasq commented on Readme.md:
```diff
@@ -304,12 +305,61 @@ Gather up the message that was sent with the http request, and save it in
 for the `on('end')` event.  Be careful when using this:  if the request has a
 body it needs to be consumed.
 
+### restiq.mw.buildReadBody(options)
```
I think the builder function writeups might work better if folded in with the mw functions,
ie describe `readBody` and `buildReadBody` together.  Certainly be easier to keep in sync.
The `options` parameter could be omitted for those builders that do not have any options.

quantumew replied:
That makes sense. Fixed this

andrasq commented on Readme.md:
```diff
@@ -304,12 +305,61 @@ Gather up the message that was sent with the http request, and save it in
+
+#### options
+
+`maxBodySize` - The maximum request body size to enforce. If this is a number, then the value specifies the number of
```
"enforce" -> "allow", "results in" -> "returns"

I think omit the size suffix writeup; I'll probably add support for it, but initially as a "stealth feature".  I'm not sure I want to commit to it formally as part of the spec.  (also, keying off string vs number seems misleading, because the value retrieved from the configuration file could still be a string, eg "100000", depending on how the configs were loaded.  Not an issue if omitted, though.)

Same elsewhere where `maxBodySize` is described.

quantumew commented:
Good deal. I will make those changes now.

andrasq commented on lib/rlib.js:
```diff
@@ -68,185 +84,237 @@ module.exports = (function() {
+            req.on('data', function() {});
+            req.on('error', function(err) {
+                req._bodyEof = true;
+                var msg = "Error atempting to dispose the request body" + (req.restiq._opts.debug ? (": " + err.stack) : "");
```
typo in "attempting"

quantumew replied:
Indeed. Fixed

andrasq commented on lib/rlib.js:
```diff
@@ -68,185 +84,237 @@ module.exports = (function() {
+
+                if (bytesReceived > options.maxBodySize) {
+                    req._bodyEof = true;
+                    readErr = new Restiq.errors.ErrorBadRequest('Error reading body, max request body size exceeded.');
```
instead of testing `readErr`, might be cleaner to write a `returnOnce` function, return the error now,
and let the read loop complete since it has to consume the rest of the response anyway.
(Edit: or forcibly abort the call, close the socket to discard the data and prevent connection hogging)
```
var _returned = false;
function returnOnce( err, ret ) {
    if (!_returned) {
        _returned = true;
        if (next) next(err, ret);
    }
}
```
Also, unrelated, I need to get rid of appendBuffer and `databuf`, since I wrote this code I've seen the
O(n^2) concat-each-buffer way be really slow.  I'll change it to the commented-out `chunks.push` version.

quantumew replied:
Good catch. If I have some time tomorrow maybe I will open another pull request and help you get that in. Looks like an easy enough change.

andrasq commented on lib/rlib.js:
```diff
@@ -68,185 +84,237 @@ module.exports = (function() {
+                        gatherChunk(chunk);
+                    }
+
+                    if (!readErr) {
```
who discards the rest of the request body if `readErr`?  (See note above about `returnOnce`)

quantumew replied:
Great point. Not sure how I missed this! Fixed

quantumew committed:
Update based on comments.

No longer parsing maxBodySize, just taking number as bytes.

quantumew commented:
Alright I got around to addressing your comments here 06dd6bb63cc9ead81a55313c5546c1d6ddb6006b - The last bullet point I don't think I fully understand, if you could elaborate on that when you take a look at the code that would be great. Thanks again sir!

andrasq reviewed (changes requested):
I think that's it for now, I know you're making changes at the same time, will look again once you're done.  Thanks!

quantumew committed:
Code review fixes, make sure body is consumed even if max size is met.

Other then that there some documentation issues.

quantumew commented:
Great feedback. Here are the fixes 4b5066273c33dfee825fc541aedff1c66f5c77f4

quantumew committed:
Added backticks around end

andrasq commented:
thanks!  I merged your pr into a new branch `merge-pr5` so I could make a few small fixes,
and merged that by hand.  Tagged as `0.7.0`.

quantumew commented:
Thanks!

quantumew commented:
Oh one last thing if you can publish this version that would be wonderful. Thanks again!

andrasq commented:
sure, published `restiq@0.7.0`
